File: DataStructure/BST/largest_bst.py
""" Given a Binary Tree, write a function that returns the size of the largest subtree which is also a Binary Search Tree (BST).
 If the complete Binary Tree is BST, then return the size of the whole tree. """
from collections import namedtuple

INT_MIN = float('-inf')
INT_MAX = float('inf')

class Node:
	def __init__(self, key):
		self.key = key
		self.left = self.right = None

node_data = namedtuple('node_data', ['is_bst', 'size', 'min', 'max', 'ans'])

def largest_bst(root):
	if not root:
		return node_data(True, 0, INT_MAX, INT_MIN, 0)

	node_data_left = largest_bst(root.left)
	node_data_right = largest_bst(root.right)

	if node_data_left.is_bst and node_data_right.is_bst and node_data_left.max < root.key < node_data_right.min :
		size = 1 + node_data_right.size + node_data_left.size
		_min = min(node_data_left.min, root.key)  # min w.r.t root for the case of no or only one child
		_max = max(node_data_right.max, root.key)
		return node_data(True, size, _min, _max, size)

	ans = max(node_data_left.size, node_data_right.size)  # pass values found earlier in subtree
	return node_data(False, 0, INT_MAX, INT_MIN, ans)

# Driver Code  
if __name__ == '__main__':  

	"""Let us construct the following Tree 
		60  
		/ \  
		65 70  
	/  
	50 """
	root = Node(60)
	root.left = Node(65)
	root.right = Node(70)
	root.left.left = Node(50)
	print("Size of the largest BST is",
					largest_bst(root).ans)

File: Cpp/largestBST.cpp
#include<bits/stdc++.h>
using namespace std;

class Node
{
	int data;
	Node* left;
	Node* right;
public:
	Node(int data):data(data),left(nullptr),right(nullptr){};
	~Node();
	
};

struct Info
{
	int min;
	int max;
	int size;
	bool isBST;
};

int largestBST(Node *root){

	if(!root) return {INT_MAX, INT_MIN,0,true};

	Info left = largestBST(root->left);
	Info right = largestBST(root->right);
	
	Info ret;

	ret._min = min(left.min, min(right.min,root->data));
	ret._max = max(left.max, max(right.max,root->data));
	ret.isBST = false;
	if(left.isBST && right.isBST && root->data > left.max && root->data < right.min){
		ret.size = 1 + left.size + right.size;
		ret.isBST = true;
	}
	else
		ret.size = max(left.size,right.size);
	return ret;
}

/* Driver program to test above functions*/
int main()
{
    /* Let us construct the following Tree
        60
       /  \
      65  70
     /
    50 */
 
    Node *root = new Node(60);
    root->left = new Node(65);
    root->right = new Node(70);
    root->left->left = new Node(50);
    printf(" Size of the largest BST is %d\n",
           largestBST(root).size);
    return 0;
}